<div style="border:solid black 2px; padding: 20px">
<div class="alert alert-block alert-info">

# Y.Music: Music Preferences among 'Springfield' and 'Shelbyville' Analysis - by awrb
    
</div>

# Table of Contents: <a id='tableofcontents'></a>

    - 1. Introduction
        - 1.1 Goal
        - 1.2 Stages
    - 2. Data Overview
        - 2.1 Data Overview Findings
    - 3. Data Pre-processing
        - 3.1 Title Writing Style
        - 3.2 Missing Values
        - 3.3 Data Duplicates
            - 3.3.1 Checking The Data Duplicates Around
        - 3.4 Data Pre-processing Findings
    - 4. Hypotheses Testing
        - 4.1 Hypothesis 1: Comparing User Behavior in Two Cities
            - 4.1.1 Hypothesis 1 Findings
        - 4.2 Hypothesis 2: Music at the Beginning and End of the Week
            - 4.2.1 Hypothesis 2 Findings
        - 4.3 Hypothesis 3: Genre Preference in Springfield and Shelbyville
            - 4.3.1 Hypothesis 3 Findings
        - 4.4 Hypotheses Findings
        - 4.5 Hypotheses Notes
    - 5. Presentation
    - 6. General Conclusions

## Introduction

        Whenever we do an analysis, we need to formulate some hypotheses that we need to test further. Sometimes, the tests we do lead us to accept the hypothesis, other times we also need to reject it. To make the right decisions in business, we must understand whether the assumptions we make are correct or not.

        In this project, we will compare the music preferences of users in the cities of 'Springfield' and 'Shelbyville'. We will study the actual Y.Music data to test the hypotheses below and compare the users behavior in these two cities.

### Goal

    In this Y.Music project, we will test three hypotheses, namely:
    
    1. User activity varies depending on the day and city.
    2. On Monday mornings, the residents of 'Springfield' and 'Shelbyville' tune into different genres. This also applies on Friday nights.
    3. Listeners in 'Springfield' and 'Shelbyville' have different preferences. In 'Springfield', the users prefer to pop music, while in 'Shelbyville' the rap music has more fans.

### Stages

    There are several steps to complete this project, namely:

    Step 1. Downloading Database and Loading The Data
    Step 2. Exploring and Rectifying the Data
    Step 3. Performing The Hypotheses Testing
    Step 4. Presenting My Project to The Viewers
    Step 5. Formulating General Conclusions

    which is divided into the following detailed stages:

    1. Data Overview
    2. Data Pre-processing
    3. Hypothesis Testing
    4. Presentation
    5. General Conclusions

    With the following description:

    1. Data Overview (Step 1. Downloading Database and Loading The Data)
    In this stage, we will download and load the dataset from the following file path:

    '/datasets/music_project_en.csv'

    Here are the data descriptions of the file path above:

    - 'userID' — user id
    - 'Track' — track title
    - 'artist' — artist name
    - 'genre' — music genres
    - 'City' — the city where the user is located
    - 'time' — the length of time the song is played
    - 'Day' — the day of the week


    2. Data Pre-processing (Step 2. Exploring and Rectifying the Data)
    In this stage, there are several things that need to be our focus, namely:

    a. Title Writing Style
    b. Missing Values
    c. Data Duplicates


    3. Hypotheses Test (Step 3. Performing The Hypotheses Testing)
    In this stage, we will perform the following Hypotheses Testing, namely:

    1. User activity varies depending on the day and city.
    2. On Monday mornings, the residents of 'Springfield' and 'Shelbyville' tune into different genres. This also applies on Friday nights.
    3. Listeners in 'Springfield' and 'Shelbyville' have different preferences. In 'Springfield', the users prefer to pop music, while in 'Shelbyville' the rap music has more fans.


    4. Presentation (Step 4. Presenting My Project to The Viewers)
    In this stage, I will present for my Project Presentation related to Music Preferences among 'Springfield' and 'Shelbyville' analysis.


    5. General Conclusions (Step 5. Formulating General Conclusions)
    In this stage, we will draw the whole of general conclusions.

[Back to Contents](#tableofcontents)

## Data Overview
    - Step 1. Downloading Database and Loading The Data

<div style="border:solid black 2px; padding: 20px"> 
    
    In this stage, we need 'Pandas' library and we will import it, as follows:
    
</div>    

In [2]:
import pandas as pd

        Now, we will load and read our file 'music_project_en.csv' from '/datasets/' folder, then we save it into 'df' variable, as follows:

In [3]:
df = pd.read_csv('/datasets/music_project_en.csv')

    We will display the first 10 table rows, as follows:

In [4]:
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

    We will get the general information about the table by a single '.dtypes' command, as follows:

In [5]:
df.dtypes

  userID    object
Track       object
artist      object
genre       object
  City      object
time        object
Day         object
dtype: object

    Temporary Findings:
    
    - There are a number of different values among each columns.
    - They all store the same data type, namely: 'object'.
    
    - Based on the documentation above, we find seven columns, namely:
        1. 'userID' — user id
        2. 'Track' — track title
        3. 'artist' — artist name
        4. 'genre' — music genres
        5. 'City' — the city where the user is located
        6. 'time' — the length of time the song is played
        7. 'Day' — the day of the week

    - We can see three cases with the style of writing column names, specifically:
        1. Some names are written in uppercase, some in lowercase.
        2. Some names use spaces.
        3. The number of column values is different. This indicates that the data we have contains missing values.
        4. Identified for data duplicates.

### Data Overview Findings

    - Each row in the table stores data related to the track of the song being played.
    - Several columns store data that describes the track itself: track title, artist, and genre. The rest stores data related to user information: their hometown, when they played the track.
    - It is clear that the datas we have are sufficient to test the hypotheses. However, we have missing values.
    - In order to proceed the analysis, we need the data pre-processing first.

[Back to Contents](#tableofcontents)

## Data Pre-processing
    - Step 2. Exploring and Rectifying the Data

<div style="border:solid black 2px; padding: 20px">   
    
        In this stage, we will rectify some cases, namely:
        1. Title Writing Style
        2. Missing Values
        3. Data Duplicates

</div>

    First, we are going to rectify format in the column headings, as follows:

### Title Writing Style

    We display column headings, as follows:

In [6]:
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

    Then, we revamp the column names according to the good writing style rules, as follows:
    1. If the name has several words, use 'snake_case'.
    2. All characters must be in lowercase.
    3. Remove spaces.

In [7]:
df = df.rename(columns={'  userID': 'user_id', 'Track': 'track', 'artist': 'artist', 'genre': 'genre', '  City  ': 'city', 'time': 'time', 'Day': 'day'})
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


    Now, we ensure the result and display the column names again, as follows:

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    63736 non-null  object
 2   artist   57512 non-null  object
 3   genre    63881 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


    Now, our title writing styles are already appropriate.
    We could continue to the next step.

[Back to Contents](#tableofcontents)

### Missing Values

    Second, we find the number of missing values in the table, as follows:

In [9]:
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

    Temporary Findings:
    
    - We found 1343 missing values on 'track' column.
    - We found 7567 missing values on 'artist' column.
    - We found 1198 missing values on 'genre' column.

    Temporary Conclusions:
    
    - Not all missing values affect the research. For example, missing values in the 'track' and 'artist' column are not significant. We can simply replace it with a clear marker. However, the missing values in the 'genre' column may influence the comparisons of musical preferences in 'Springfield' and 'Shelbyville'.
    - In real life, it is very useful to learn the reasons why the data is missing and try to fix it. Unfortunately, we have not had that opportunity in this project. Therefore, we have to:
    
        1. Filling in missing values with markers.
        2. Evaluating how many missing values can affect our calculations.

    In this session, we are going to replace missing values in 'track', 'artist' and 'genre' column with string 'unknown'. To do this, we will create 'list' of 'columns_to_replace', apply 'loop' with 'for' at that 'list', and replace the missing values in each columns, as follows:

In [10]:
columns_to_replace = ['track', 'artist', 'genre']
for col in columns_to_replace:
    df[col].fillna('unknown', inplace=True)
    
print(columns_to_replace)

['track', 'artist', 'genre']


    We will ensure no more tables containing missing values and recalculate the missing values, as follows:

In [11]:
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

    Now, our dataset is clean from missing values.
    We could continue to the next step.

[Back to Contents](#tableofcontents)

### Data Duplicates

    Third, we will find the number of explicit data duplicates in the table by using a single '.duplicated().sum()' command, as follows:

In [12]:
df.duplicated().sum()

3826

    Temporary Findings:
    - We found 3826 data duplicates in our dataset.

    Then, we are going to remove data duplicates above, as follows:

In [13]:
df = df.drop_duplicates()

    Now, we ensure the datas, as follows:

In [14]:
df.duplicated().sum()

0

    We calculate the explicit data duplicates again to ensure we have removed them all, as follows:

In [15]:
df['genre'].isna().sum()

0

    Now, our dataset is clean from data duplicates. Our dataset is ready to be processed and analyzed.
    We could continue to the next step.

#### Checking The Data Duplicates Around

    In this session, we will remove the implicit data duplicates in the 'genre' column. For example, genre names can be written in different ways. Errors like these will also affect our results.

    Displaying 'list' that containing the unique genre name, then sorting that 'list' alphabetically. For doing this, we have to:
    
    1. Retrieving the DataFrame column in question.
    2. Applying a sort method to that column.
    3. For a sorted column, summoning a method that returns all of the unique values columns.

In [16]:
sorted(df['genre'].unique())

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

    Taking a look carefully for the 'list' to find the implicit data duplicates of the 'hiphop' genre. Those data duplicates could be an incorrectly written name or an alternative name from the same genre.
    
    We will find the following implicit data duplicates, as follows:
    - hip
    - hop
    - hip-hop
    
    To remove them, we will use the 'replace_wrong_genres()' function by two parameters, those are:
    - 'wrong_genres' = — The 'list' of the data duplicates we want to replace
    - 'correct_genre' = — The 'string' with the correct value
    
    That function must correct the name in the 'genre' column from 'df' table by replacing each values of 'wrong_genres' list with the value of 'correct_genre'.

    Now, we start for the function, as follows:

In [17]:
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)
        
duplicates = ['hip', 'hop', 'hip-hop']
genre = 'hiphop'
replace_wrong_genres(duplicates, genre)
print(genre)

hiphop


    In this session, we will summon the 'replace_wrong_genres()' function and forward the arguments to that function, so it could remove implicit data duplicates ('hip', 'hop', and 'hip-hop'), then replace them to 'hiphop'.

    Entering a function that replaces implicit data duplicates, as follows:

In [18]:
df['genre'].unique()

array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', 'unknown', 'alternative', 'children', 'rnb',
       'hiphop', 'jazz', 'postrock', 'latin', 'classical', 'metal',
       'reggae', 'triphop', 'blues', 'instrumental', 'rusrock', 'dnb',
       'türk', 'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock

    Ensuring that data duplicates values have been removed. Displaying 'list' of unique values from 'genre' column and we will check for implicit data duplicates, as follows:

In [47]:
sorted(df['genre'].unique())

['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

    Now, we have checked for The Data Duplicates Around.
    We go to the next step.

[Back to Contents](#tableofcontents)

### Data Pre-processing Findings

    - We have detected three cases in our data:
    1. Incorrect title writing style
    2. Missing values
    3. Explicit and implicit data duplicates
    
    - Column headings have now been cleaned up to make table processing easier.
    - All missing values have been replaced by 'unknown'. However, we still have to see if the missing values in the 'genre' column whether will affect our calculations.
    - The absence of data duplicates will make the results we get more precise and easier to understand.
    
    
    Now, we could move on to hypotheses testing.

[Back to Contents](#tableofcontents)

## Hypotheses Testing
    - Step 3. Performing The Hypotheses Testing

<div style="border:solid black 2px; padding: 20px"> 
    
    In this stage, we have Three Hypotheses for our concern in testing method, those are:

    - Hypothesis 1: Comparing User Behavior in Two Cities
    - Hypothesis 2: Music at the Beginning and End of the Week
    - Hypothesis 3: Genre Preference in Springfield and Shelbyville
    
</div>

### Hypothesis 1: Comparing User Behavior in Two Cities
    - User activity varies depending on the day and city

    According to the first hypothesis, the users from 'Springfield' and 'Shelbyville' have different behavior in listening to music. This test uses the data taken from three days of the week: Monday, Wednesday, and Friday.
    
    - Dividing the users into groups by city.
    - Comparing how many tracks each groups played on Monday, Wednesday, and Friday.

    We will do each calculations separately and evaluate user activity in each cities. Then, we make group of the data by city and find the number of tracks playing in each groups.

    Calculating the tracks of songs played in each cities, as follows:

In [54]:
df.groupby('city')['track'].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

    Temporary Findings:
    
    - We found 18512 tracks of songs played in 'Shelbyville' city.
    - We found 42741 tracks of songs played in 'Springfield' city.

    The users from 'Springfield' played more tracks than the users from 'Shelbyville'. However, this does not imply that 'Springfield' residents listen to music more often. That city is bigger, and there are more users.
    
    
    Now, we are going to make group of the data by day and find the number of songs tracks played on Monday, Wednesday and Friday.

    Calculating the tracks of songs played on each days, as follows:

In [57]:
df.groupby('day')['track'].count()

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

    Temporary Findings:
    
    - We found 21840 tracks of songs played on 'Friday'.
    - We found 21354 tracks of songs played on 'Monday'.
    - We found 18059 tracks of songs played on 'Wednesday'.

    Wednesday is the quietest day overall. However if we consider the two cities separately, we might come to a different conclusion.

     We have seen how grouping by city or day works. Now, we are going to write a function that will segment the data by city and day.
     
     We make 'number_tracks()' function to calculate the number of songs tracks played for certain day and city. The function will take two parameters, namely:
     
     - the days name of the week
     - city name
     
     In the function we created, we will use a variable to store the rows from the original table, where:
     
     - The 'day' column value is the same as the 'day' parameter
     - The 'city' column value is the same as the 'city' parameter
     
     We will apply sequential filtering with logical indexing. Afterwards, we calculate the value of 'user_id' column in the resulting table. Then, we save the result into a new variable and return this variable from the function.

    Creating 'number_tracks()' function, as follows:

In [62]:
def number_tracks(day, city):
    df_filter = df[(df['day'] == day) & (df['city'] == city)]
    track_list_count = df_filter['user_id'].count()
    return track_list_count

    Now, we summon the 'number_tracks()' function in six times and change the parameter value on each summons, with the result that we could retrieve the datas for both cities for each days (Monday, Wednesday, and Friday).

    First, the number of songs played at 'Springfield' on Monday, as follows:

In [75]:
number_tracks('Monday', 'Springfield')

15740

    Second, the number of songs played in 'Shelbyville' on Monday, as follows:

In [76]:
number_tracks('Monday', 'Shelbyville')

5614

    Third, the number of songs played at 'Springfield' on Wednesday, as follows:

In [77]:
number_tracks('Wednesday', 'Springfield')

11056

    Fourth, the number of songs played in 'Shelbyville' on Wednesday, as follows:

In [78]:
number_tracks('Wednesday', 'Shelbyville')

7003

    Fifth, the number of songs played at 'Springfield' on Friday, as follows:

In [79]:
number_tracks('Friday', 'Springfield')

15945

    Sixth, the number of songs played in 'Shelbyville' on Friday, as follows:

In [80]:
number_tracks('Friday', 'Shelbyville')

5895

    Temporary Findings:
    
    - We have 15740 number of songs played at 'Springfield' on Monday.
    - We have 5614 number of songs played in 'Shelbyville' on Monday.
    - We have 11056 number of songs played at 'Springfield' on Wednesday.
    - We have 7003 number of songs played in 'Shelbyville' on Wednesday.
    - We have 15945 number of songs played at 'Springfield' on Friday.
    - We have 5895 number of songs played in 'Shelbyville' on Friday.

    Now, we will use 'pd.DataFrame' to create a table, that is:
    
    - The column names are: ['city', 'monday', 'wednesday', 'friday']
    - The data is the result of 'number_tracks()' function

    This is table by the results, as follows:

In [84]:
column = ['city', 'monday', 'wednesday', 'friday']
path = [
    ['Springfield', number_tracks('Monday', 'Springfield'), number_tracks('Wednesday', 'Springfield'), number_tracks('Friday', 'Springfield')],
    ['Shelbyville', number_tracks('Monday', 'Shelbyville'), number_tracks('Wednesday', 'Shelbyville'), number_tracks('Friday', 'Shelbyville')]
]

table = pd.DataFrame(data=path, columns=column)
table

city  monday  wednesday  friday
0  Springfield   15740      11056   15945
1  Shelbyville    5614       7003    5895

#### Hypothesis 1 Findings

    Conclusions:
    
    - The data we get reveals differences in user behavior:
    
    1. In the city of 'Springfield', the track play peaked is on Monday and Friday, while Wednesday saw a decline in activity.
    2. In the city of 'Shelbyville', on the other hand, the users listened to more music on Wednesdays.
    
    - Less users activity on Mondays and Fridays.

[Back to Contents](#tableofcontents)

### Hypothesis 2: Music at the Beginning and End of the Week
    - On Monday mornings, the residents of 'Springfield' and 'Shelbyville' tune into different genres. This also applies on Friday nights

    According to the second hypothesis, on Monday mornings and Friday evenings, the 'Springfield' residents listen to a different genre of music than the people of 'Shelbyville' enjoy.

    Getting the table (ensuring our join table name matches the DataFrame given in the two code blocks below):
    
    - For 'Springfield' — spr_general
    - For 'Shelbyville' — shel_general

    Now we will get the 'spr_general' table from 'df' row, where the value of 'city' column is 'Springfield', as follows:

In [91]:
spr_general = df[df['city'] == 'Springfield']
spr_general

user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE                       L’estate              Julia Dalia   
...         ...                            ...                      ...   
65073  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
65074  729CBB09                        My Name                   McLean   
65076  C5E3A0D5                      Jalopiina                  unknown   
65077  321D0506                  Freight Train            Chas McDevitt   
65078  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre         city      time        day  
1              rock  Springfield  14:07:09     Friday  
4             dance  Springfield  08:34:34     Monday  
6             dance  Springfield  13:00:07  Wednesday  
7             dance  Springfield  20:47:49  Wednesday  
8            ruspop  Springfield  09:17:40     Friday  
...             ...          ...       ...        ...  
65073  extrememetal  Springfield  21:07:12     Monday  
65074           rnb  Springfield  13:32:28  Wednesday  
65076    industrial  Springfield  20:09:26     Friday  
65077          rock  Springfield  21:43:59     Friday  
65078       country  Springfield  21:59:46     Friday  

[42741 rows x 7 columns]

    Then, we will get 'shell_general' table from 'df' row, where the value of 'city' column is 'Shelbyville', as follows:

In [92]:
shel_general = df[df['city'] == 'Shelbyville']
shel_general

user_id                              track              artist  \
0      FFB692EC                  Kamigata To Boots    The Mass Missile   
2        20EC38                  Funiculì funiculà         Mario Lanza   
3      A3DD03C9              Dragons in the Sunset          Fire + Ice   
5      842029A1                             Chains            Obladaet   
9      E772D5C0                          Pessimist             unknown   
...         ...                                ...                 ...   
65063  D94F810B        Theme from the Walking Dead  Proyecto Halloween   
65064  BC8EC5CF       Red Lips: Gta (Rover Rework)               Rover   
65065  29E04611                       Bre Petrunko       Perunika Trio   
65066  1B91C621             (Hello) Cloud Mountain     sleepmakeswaves   
65075  D08D4A55  Maybe One Day (feat. Black Spade)         Blu & Exile   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
5          rusrap  Shelbyville  13:09:41     Friday  
9           dance  Shelbyville  21:20:49  Wednesday  
...           ...          ...       ...        ...  
65063        film  Shelbyville  21:14:40     Monday  
65064  electronic  Shelbyville  21:06:50     Monday  
65065       world  Shelbyville  13:56:00     Monday  
65066    postrock  Shelbyville  09:22:13     Monday  
65075      hiphop  Shelbyville  10:00:00     Monday  

[18512 rows x 7 columns]

    Furthermore, we write a 'genre_weekday()' function with four parameters:
    
    - A table for data
    - Name of the day
    - First timestamp, in 'hh:mm' format
    - Last timestamp, in 'hh:mm' format
    
    
    That function should return information about the 15 most popular genres on a certain day in the period among two timestamps.

    Declaring the 'genre_weekday()' function with parameters day=, time1=, and time2=. That function must provides information on the most popular genres on a certain day and time, as follows:

In [98]:
def genre_weekday(df,day,time2,time1):
    genre_df = df[df['day'] == day]
    genre_df = genre_df[genre_df['time'] < time2]
    genre_df = genre_df[genre_df['time'] > time1]
    genre_df_grouped = genre_df.groupby('genre')['user_id'].count()
    genre_df_sorted = genre_df_grouped.sort_values(ascending=False) 
    return genre_df_sorted[:15]

    Now, we will compare the results of the genre_weekday() function for 'Springfield' and 'Shelbyville' on a Monday morning (7:00 a.m. to 11:00 a.m.) and a Friday night (from 17:00 p.m. to 23:00 p.m.), as follows:

    1. Summoning the function for Monday morning in 'Springfield' by using 'spr_general' instead of 'df' table.

In [101]:
genre_weekday(spr_general,'Monday','11.00','07.00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
classical      157
metal          120
jazz           100
folk            97
soundtrack      95
Name: user_id, dtype: int64

    2. Summoning the function for Monday morning in 'Shelbyville' by using 'shell_general' instead of 'df' table.

In [103]:
genre_weekday(shel_general,'Monday','11.00','07.00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
world           36
rap             32
soundtrack      31
rnb             27
metal           27
Name: user_id, dtype: int64

    3. Summoning the function for Friday night in 'Springfield' by using 'spr_general' instead of 'df' table.

In [106]:
genre_weekday(spr_general,'Friday','23.00','17.00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
classical      163
alternative    163
rusrap         142
jazz           111
unknown        110
soundtrack     105
rnb             90
metal           88
Name: user_id, dtype: int64

    4. Summoning the function for Friday night in 'Shelbyville' by using 'shel_general' instead of 'df' table.

In [108]:
genre_weekday(shel_general,'Friday','23.00','17.00')

genre
pop            256
rock           216
electronic     216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
unknown         47
ruspop          47
soundtrack      40
metal           39
rap             36
Name: user_id, dtype: int64

    In this session, we have already compared the results of the genre_weekday() function for 'Springfield' and 'Shelbyville' on a Monday morning (7:00 a.m. to 11:00 a.m.) and a Friday night (from 17:00 p.m. to 23:00 p.m.).
    
    Now, we go to the next step.

#### Hypothesis 2 Findings

    Conclusions:
    
    - After comparing the top 15 genres on Monday morning, we can draw the following conclusions:
    
    1. The users from 'Springfield' and 'Shelbyville' listened to the music of the same genre. The top five genres from both cities are the same, only the rock and electronic genres switch places.
    2. In 'Springfield', the number of missing values is surprisingly large, so the value 'unknown' is 10th. This means that the missing values encompass for a sizeable proportion of the data, so this fact could serve as a basis for questioning the reliability of our conclusions.
    
    - For Friday night, the situation is also similar. Individual genres vary quite a bit, but overall, the top 15 genres for both cities are the same.
    - Thus, the second hypothesis is partially proven correct of:
    
    1. The users listen to the same music at the beginning and end of the week.
    2. There is no notable difference among 'Springfield' and 'Shelbyville'. In both cities, pop is the most popular genre.
    
    - However, the significance number of missing values makes this result questionable. In 'Springfield', there are so many missing values that influence our top 15 genre results. If we did not have these missing values, the results might have been different.

[Back to Contents](#tableofcontents)

### Hypothesis 3: Genre Preference in Springfield and Shelbyville
    - Listeners in 'Springfield' and 'Shelbyville' have different preferences. In 'Springfield', the users prefer to pop music, while in 'Shelbyville' the rap music has more fans

    Hypothesis: 'Shelbyville' residents loves rap music. 'Springfield' residents prefers pop.

    Grouping the 'spr_general' table by genre and find the number of played tracks for each genres by the 'count()' method. Then, sorting the results in descending order and save them to 'spr_genres' table.

    We will do for some steps below:
    
    - in one row: grouping 'spr_general' table by column 'genre'
    - calculating the value of 'genre' column by 'count()' in grouping
    - sorting the resulting Series in descending order, then saving the result to 'spr_genres'

In [110]:
spr_genres = spr_general.groupby('genre')['track'].count()
spr_genres = spr_genres.sort_values(ascending=False)
spr_genres

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
               ... 
metalcore         1
marschmusik       1
malaysian         1
lovers            1
ïîï               1
Name: track, Length: 250, dtype: int64

    Now, we display for the first 10 rows of 'spr_genres', as follows:

In [111]:
spr_genres.head(10)

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: track, dtype: int64

    Now, we do the same with the data from 'Shelbyville'.
    
    Grouping the 'shell_general' table by genre and find the number of played tracks for each genres. Then, sorting the results in descending order and saving the results to the 'shel_genres' table.

    We will do for some steps below:
    
    - in one row: grouping 'shell_general' table by 'genre' column
    - calculating the value of 'genre' column in grouping by using count()
    - sorting the resulting Series in descending order and saving them to 'shell_genres'

In [113]:
shel_genres = shel_general.groupby('genre')['track'].count()
shel_genres = shel_genres.sort_values(ascending=False)
shel_genres

genre
pop           2431
dance         1932
rock          1879
electronic    1736
hiphop         960
              ... 
mandopop         1
leftfield        1
laiko            1
jungle           1
worldbeat        1
Name: track, Length: 202, dtype: int64

    Now, we display for the first 10 rows of 'shel_genres', as follows:

In [115]:
shel_genres.head(10)

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: track, dtype: int64

#### Hypothesis 3 Findings

    Conclusions:
    
    - This hypothesis has been partially proven that:
    
    1. Pop music is the most popular genre in 'Springfield', as one would expect.
    2. However, pop music was equally popular in both 'Springfield' and 'Shelbyville', and rap music did not make the top 5 genre lists for either cities.

[Back to Contents](#tableofcontents)

### Hypotheses Findings

    - We have tested the following three hypotheses:
    
    1. The users activity in 'Springfield' and 'Shelbyville' depend on the day of the week, although these two cities vary in a number of ways.
    2. On Monday mornings, 'Springfield' and 'Shelbyville' residents tune into different genres. This also applies on Friday nights.
    3. 'Springfield' and 'Shelbyville' listeners have different preferences. In both 'Springfield' and 'Shelbyville', they preferred pop music.
    
    - After analyzing the available data, we could conclude that:
    
    1. The users activity in 'Springfield' and 'Shelbyville' depend on the day, even if the city is different.
    Thus, the first hypothesis can be fully accepted.
    2. Musical preferences did not vary significantly throughout the week in 'Springfield' and 'Shelbyville'. We could see a small difference in the order on Monday, but:
    In both 'Springfield' and 'Shelbyville', the users mostly listen to pop music.
    
    Therefore, we could not accept this hypothesis. It is also important to remember that the results obtained could have been different had we not had the missing values.
    
    3. It turns out that the music preferences of users from 'Springfield' and 'Shelbyville' are very similar.
    The third hypothesis is rejected. If there are indeed differences in preference, unfortunately we could not know this from this data.

[Back to Contents](#tableofcontents)

### Hypotheses Notes 

    Notes:
    
    - The research involves statistical hypothesis testing, which is certainly more precise and more quantitative. Also note that we could not always draw the conclusions about an entire city based on data from just one source.

[Back to Contents](#tableofcontents)

## Presentation
    - Step 4. Presenting My Project to The Viewers

<div style="border:solid black 2px; padding: 20px">
    
    This is my presentation of 'Y.Music: Music Preferences among 'Springfield' and 'Shelbyville' Analysis' project by me, awrb.

</div>

Presentation Link: https://drive.google.com/file/d/1bhH9e-nIcudI0ukHi_sDpEd9Kv70HcYa/view?usp=sharing

[Back to Contents](#tableofcontents)

## General Conclusions

<div style="border:solid black 2px; padding: 20px">
    
# For Data Overview
    
    1. Temporary Findings:
    
        - There are a number of different values among each columns.
        - They all store the same data type, namely: 'object'.
    
        - Based on the documentation above, we find seven columns, namely:
            1. 'userID' — user id
            2. 'Track' — track title
            3. 'artist' — artist name
            4. 'genre' — music genres
            5. 'City' — the city where the user is located
            6. 'time' — the length of time the song is played
            7. 'Day' — the day of the week

        - We can see three cases with the style of writing column names, specifically:
            1. Some names are written in uppercase, some in lowercase.
            2. Some names use spaces.
            3. The number of column values is different. This indicates that the data we have contains missing values.
            4. Identified for data duplicate.
    
    
    2. Data Overview Findings
    
        - Each row in the table stores data related to the track of the song being played.
        - Several columns store data that describes the track itself: track title, artist, and genre. The rest stores data related to user information: their hometown, when they played the track.
        - It is clear that the datas we have are sufficient to test the hypothesis. However, we have missing values.
        - In order to proceed the analysis, we need the data pre-processing first.
    
</div>

<div style="border:solid black 2px; padding: 20px">
    
# For Data Pre-processing
    
    A. Title Writing Style
    
    1. For rectifying the Title Writing Style, we revamp the column names according to the good writing style rules, as follows:
        - If the name has several words, use 'snake_case'.
        - All characters must be in lowercase.
        - Remove spaces.

    
    B. Missing Values
    
    1. Temporary Findings:
    
        - We found 1343 missing values on 'track' column.
        - We found 7567 missing values on 'artist' column.
        - We found 1198 missing values on 'genre' column.
    
    2. Temporary Conclusions:
    
        - Not all missing values affect the research. For example, missing values in the 'track' and 'artist' column are not significant. We can simply replace it with a clear marker. However, the missing values in the 'genre' column may influence the comparisons of musical preferences in 'Springfield' and 'Shelbyville'.
        - In real life, it is very useful to learn the reasons why the data is missing and try to fix it. Unfortunately, we have not had that opportunity in this project. Therefore, we have to:
    
            1. Filling in missing values with markers.
            2. Evaluating how many missing values can affect our calculations.
    
    
    C. Data Duplicate 
    
    3. Temporary Findings:
    
        - We found 3826 data duplicates in our dataset.
    

    D. Data Pre-processing Findings
    
    - We have detected three problems in our data:
        1. Incorrect title writing style
        2. Missing values
        3. Explicit and implicit data duplicates

    - Column headings have now been cleaned up to make table processing easier.
    - All missing values have been replaced by 'unknown'. However, we still have to see if the missing values in the 'genre' column whether will affect our calculations.
    - The absence of data duplicates will make the results we get more precise and easier to understand.
    
</div>

<div style="border:solid black 2px; padding: 20px">
    
# For Hypothesis Testing
    
    A. Hypothesis 1: Comparing User Behavior in Two Cities
    
    1. Temporary Findings:

        - We found 18512 tracks of songs played in 'Shelbyville' city.
        - We found 42741 tracks of songs played in 'Springfield' city.
    
    2. Temporary Findings:
    
        - We found 21840 tracks of songs played on 'Friday'.
        - We found 21354 tracks of songs played on 'Monday'.
        - We found 18059 tracks of songs played on 'Wednesday'.

    Wednesday is the quietest day overall. However if we consider the two cities separately, we might come to a different conclusion.
    
    3. Temporary Findings:
    
        - We have 15740 number of songs played at 'Springfield' on Monday.
        - We have 5614 number of songs played in 'Shelbyville' on Monday.
        - We have 11056 number of songs played at 'Springfield' on Wednesday.
        - We have 7003 number of songs played in 'Shelbyville' on Wednesday.
        - We have 15945 number of songs played at 'Springfield' on Friday.
        - We have 5895 number of songs played in 'Shelbyville' on Friday.
    
    4. Hypothesis 1 Findings
    
    Conclusions:

    - The data we get reveals differences in user behavior:

        1. In the city of 'Springfield', the track play peaked is on Monday and Friday, while Wednesday saw a decline in activity.
    2. In the city of 'Shelbyville', on the other hand, the users listened to more music on Wednesdays.

    - Less users activity on Mondays and Fridays.
    
    
    B. Hypothesis 2: Music at the Beginning and End of the Week
    
    1. we have already compared the results of the genre_weekday() function for 'Springfield' and 'Shelbyville' on a Monday morning (7:00 a.m. to 11:00 a.m.) and a Friday night (from 17:00 p.m. to 23:00 p.m.).

    2. Hypothesis 2 Findings
    
    Conclusions:

    - After comparing the top 15 genres on Monday morning, we can draw the following conclusions:

        1. The users from 'Springfield' and 'Shelbyville' listened to the music of the same genre. The top five genres from both cities are the same, only the rock and electronic genres switch places.
        2. In 'Springfield', the number of missing values is surprisingly large, so the value 'unknown' is 10th. This means that the missing values encompass for a sizeable proportion of the data, so this fact could serve as a basis for questioning the reliability of our conclusions.

    - For Friday night, the situation is also similar. Individual genres vary quite a bit, but overall, the top 15 genres for both cities are the same.
    - Thus, the second hypothesis is partially proven correct of:

        1. The users listen to the same music at the beginning and end of the week.
        2. There is no notable difference among 'Springfield' and 'Shelbyville'. In both cities, pop is the most popular genre.

    - However, the significance number of missing values makes this result questionable. In 'Springfield', there are so many missing values that influence our top 15 genre results. If we did not have these missing values, the results might have been different.
    
    
    C. Hypothesis 3: Genre Preference in Springfield and Shelbyville
    
    1. Hypothesis 3 Findings
    
    Conclusions:

    - This hypothesis has been partially proven that:

        1. Pop music is the most popular genre in 'Springfield', as one would expect.
        2. However, pop music was equally popular in both 'Springfield' and 'Shelbyville', and rap music did not make the top 5 genre lists for either cities.
    
    
    D. Hypothesis Findings
    
    - We have tested the following three hypotheses:

        1. The users activity in 'Springfield' and 'Shelbyville' depend on the day of the week, although these two cities vary in a number of ways.
        2. On Monday mornings, 'Springfield' and 'Shelbyville' residents tune into different genres. This also applies on Friday nights.
        3. 'Springfield' and 'Shelbyville' listeners have different preferences. In both 'Springfield' and 'Shelbyville', they preferred pop music.

    - After analyzing the available data, we could conclude that:

        1. The users activity in 'Springfield' and 'Shelbyville' depend on the day, even if the city is different.
    Thus, the first hypothesis can be fully accepted.
        2. Musical preferences did not vary significantly throughout the week in 'Springfield' and 'Shelbyville'. We could see a small difference in the order on Monday, but:
    In both 'Springfield' and 'Shelbyville', the users mostly listen to pop music.

    Therefore, we could not accept this hypothesis. It is also important to remember that the results obtained could have been different had we not had the missing values.

        3. It turns out that the music preferences of users from 'Springfield' and 'Shelbyville' are very similar.
    The third hypothesis is rejected. If there are indeed differences in preference, unfortunately we could not know this from this data.
    
    
    E. Hypothesis Notes
    
    Notes:

    - The research involves statistical hypothesis testing, which is certainly more precise and more quantitative. Also note that we could not always draw the conclusions about an entire city based on data from just one source.
    
</div>

<div style="border:solid black 2px; padding: 20px">
    
# Presentation
    
    Presentation Link: https://drive.google.com/file/d/1bhH9e-nIcudI0ukHi_sDpEd9Kv70HcYa/view?usp=sharing
    
</div>

[Back to Contents](#tableofcontents)